File: mobile/ios/Domofon/Events/EventsViewModel.swift
import Foundation

@MainActor
final class EventsViewModel: ObservableObject {
    @Published var items: [EventDto] = []
    @Published var nextCursor: String?
    @Published var loading = false
    @Published var error: String?

    func refresh() {
        Task {
            loading = true
            error = nil
            do {
                let page = try await APIClient.shared.events()
                items = page.items
                nextCursor = page.nextCursor
            } catch {
                self.error = "Не удалось загрузить события"
            }
            loading = false
        }
    }

    func loadMore() {
        guard let cursor = nextCursor else { return }
        Task {
            do {
                let page = try await APIClient.shared.events(cursor: cursor)
                items += page.items
                nextCursor = page.nextCursor
            } catch {
                // pagination errors are non-fatal
            }
        }
    }
}

File: mobile/android/app/src/main/java/ru/domofon/app/events/EventsViewModel.kt
package ru.domofon.app.events

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import ru.domofon.app.network.DomofonApi
import ru.domofon.app.network.EventDto
import javax.inject.Inject

data class EventsUiState(
    val items: List<EventDto> = emptyList(),
    val nextCursor: String? = null,
    val loading: Boolean = false,
    val error: String? = null
)

@HiltViewModel
class EventsViewModel @Inject constructor(
    private val api: DomofonApi
) : ViewModel() {
    private val _state = MutableStateFlow(EventsUiState())
    val state: StateFlow<EventsUiState> = _state.asStateFlow()

    init {
        refresh()
    }

    fun refresh() {
        viewModelScope.launch {
            _state.update { it.copy(loading = true, error = null) }
            runCatching { api.events() }
                .onSuccess { page ->
                    _state.update {
                        it.copy(loading = false, items = page.items, nextCursor = page.nextCursor)
                    }
                }
                .onFailure { e ->
                    _state.update { it.copy(loading = false, error = e.message ?: "Ошибка сети") }
                }
        }
    }

    fun loadMore() {
        val cursor = _state.value.nextCursor ?: return
        viewModelScope.launch {
            runCatching { api.events(cursor) }
                .onSuccess { page ->
                    _state.update {
                        it.copy(items = it.items + page.items, nextCursor = page.nextCursor)
                    }
                }
        }
    }
}
